src/java/org/apache/cassandra/utils/ByteComparable.java

```diff
@@ -0,0 +1,165 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.utils;
+
+import java.nio.ByteBuffer;
+
+import static org.apache.cassandra.utils.ByteSource.END_OF_STREAM;
+
+/**
+ * Interface indicating a value can be represented/identified by a comparable {@link ByteSource}.
+ */
+public interface ByteComparable
+{
+    /**
+     * Returns a source that generates the byte-comparable representation of the value byte by byte.
+     */
+    ByteSource asComparableBytes(Version version);
+
+    default ByteSource.Peekable asPeekableBytes(Version version)
+    {
+        return ByteSource.peekable(asComparableBytes(version));
+    }
+
+    enum Version
+    {
+        OSS41,  // CASSANDRA 4.1 encoding, used in trie-based indices
```

Comment:
We do need the two versions, at least for the DSE7 (as opposed to OSS C* 4+) version of this, so that we can read DSE6 files, as well as SAI files, which AFAIK use DSE68 encoding.

Even in the OSS patch, I think it makes sense to keep the two versions, together with their DSE naming, in order to leave the final version choice to be determined later (It would be great if C* 4 could read DSE6 files, but TBH I don't think that's realistic).

Reply:
So is that mandatory to have that original DSE naming? Do we use the actual enum names rather than their ordinal numbers? I was thinking that It could be safe to name them `LEGACY` and `OSS41`

Reply:
I think the original names are a bit of context that will be a little helpful to reviewers and contributors and something we don't need to hide. But I don't have strong feelings, if you definitely prefer to rename them, do so.